File: lib/oauth-mcp-client.ts
import { MCPClient } from '@modelcontextprotocol/sdk';
import * as fs from 'fs';
import * as path from 'path';

export class OAuthMCPClient extends MCPClient {
  private cacheDir: string;
  private clientIdFile: string;
  private tokenFile: string;

  constructor(options: { cacheDir?: string } = {}) {
    super();
    this.cacheDir = options.cacheDir || path.join(process.cwd(), '.mcp-cache');
    this.clientIdFile = path.join(this.cacheDir, 'client_id.json');
    this.tokenFile = path.join(this.cacheDir, 'tokens.json');
  }

  /**
   * Handle OAuth completion with error recovery for invalid client IDs
   */
  async completeOAuth(code: string, state: string): Promise<void> {
    try {
      await super.completeOAuth(code, state);
    } catch (error) {
      if (this.isInvalidClientIdError(error)) {
        console.warn('Invalid client ID detected. Resetting client state and re-registering...');
        await this.resetClientState();
        await this.registerClient();
        // Retry the OAuth completion with new client ID
        await super.completeOAuth(code, state);
      } else {
        throw error;
      }
    }
  }

  /**
   * Check if error is due to invalid/expired client ID
   */
  private isInvalidClientIdError(error: any): boolean {
    const errorMessage = error.message || error.toString();
    return errorMessage.includes('client_id') &&
           (errorMessage.includes('invalid') ||
            errorMessage.includes('expired') ||
            errorMessage.includes('not found'));
  }

  /**
   * Reset client state by clearing cached files
   */
  async resetClientState(): Promise<void> {
    try {
      // Ensure cache directory exists
      await fs.promises.mkdir(this.cacheDir, { recursive: true });

      // Remove cached client ID and tokens
      const filesToRemove = [this.clientIdFile, this.tokenFile];
      for (const file of filesToRemove) {
        try {
          await fs.promises.unlink(file);
        } catch (err) {
          // Ignore if file doesn't exist
        }
      }

      console.log('Client state reset successfully');
    } catch (error) {
      console.error('Failed to reset client state:', error);
      throw error;
    }
  }

  /**
   * Register client and store new client ID
   */
  async registerClient(): Promise<void> {
    try {
      // Clear any existing client ID
      this.clientId = undefined;

      // Call the registration endpoint
      const response = await this.makeRequest('/register', {
        method: 'POST',
        body: {
          client_name: 'MCP OAuth Client',
          redirect_uris: [this.redirectUri],
          scope: this.scopes.join(' ')
        }
      });

      this.clientId = response.client_id;

      // Cache the new client ID
      await fs.promises.writeFile(
        this.clientIdFile,
        JSON.stringify({
          client_id: this.clientId,
          created_at: new Date().toISOString()
        }, null, 2)
      );

      console.log('Client registered successfully with ID:', this.clientId);
    } catch (error) {
      console.error('Failed to register client:', error);
      throw error;
    }
  }

  /**
   * Initialize client with automatic state recovery
   */
  async initialize(): Promise<void> {
    try {
      // Try to load cached client ID
      await this.loadCachedState();

      // Initialize the parent client
      await super.initialize();
    } catch (error) {
      if (this.isInvalidClientIdError(error)) {
        console.warn('Cached client ID invalid. Resetting and re-registering...');
        await this.resetClientState();
        await this.registerClient();
        await super.initialize();
      } else {
        throw error;
      }
    }
  }

  /**
   * Load cached client state
   */
  private async loadCachedState(): Promise<void> {
    try {
      const clientIdData = await fs.promises.readFile(this.clientIdFile, 'utf-8');
      const { client_id } = JSON.parse(clientIdData);
      this.clientId = client_id;

      // Optional: check if client ID is expired (implement based on your server's TTL)
      // const { created_at } = JSON.parse(clientIdData);
      // const age = Date.now() - new Date(created_at).getTime();
      // if (age > 24 * 60 * 60 * 1000) { // 24 hours
      //   throw new Error('Client ID expired');
      // }

    } catch (error) {
      // No cached state or invalid, will register new client
    }
  }
}

// Usage example:
async function createMCPClient() {
  const client = new OAuthMCPClient({
    cacheDir: './.mcp-cache' // Optional custom cache directory
  });

  try {
    await client.initialize();
    console.log('MCP client initialized successfully');
    return client;
  } catch (error) {
    console.error('Failed to initialize MCP client:', error);
    throw error;
  }
}

export default OAuthMCPClient;

File: lib/oauth_mcp_client.py
import os
import json
import asyncio
from pathlib import Path
from typing import Optional, Dict, Any
from mcp import ClientSession, stdio_client

class OAuthMCPClient:
    def __init__(self, cache_dir: Optional[str] = None):
        self.cache_dir = Path(cache_dir or ".mcp-cache")
        self.client_id_file = self.cache_dir / "client_id.json"
        self.token_file = self.cache_dir / "tokens.json"
        self.client_id: Optional[str] = None
        self.session: Optional[ClientSession] = None

    async def complete_oauth(self, code: str, state: str) -> None:
        """Handle OAuth completion with error recovery for invalid client IDs"""
        try:
            if self.session:
                # Use the session's OAuth completion method
                await self.session.complete_oauth(code, state)
        except Exception as error:
            if self._is_invalid_client_id_error(error):
                print("Invalid client ID detected. Resetting client state and re-registering...")
                await self._reset_client_state()
                await self._register_client()
                # Retry with new client ID
                await self.session.complete_oauth(code, state)
            else:
                raise error

    def _is_invalid_client_id_error(self, error: Exception) -> bool:
        """Check if error is due to invalid/expired client ID"""
        error_message = str(error).lower()
        return ('client_id' in error_message and
                any(keyword in error_message for keyword in ['invalid', 'expired', 'not found']))

    async def _reset_client_state(self) -> None:
        """Reset client state by clearing cached files"""
        try:
            # Ensure cache directory exists
            self.cache_dir.mkdir(parents=True, exist_ok=True)

            # Remove cached files
            files_to_remove = [self.client_id_file, self.token_file]
            for file_path in files_to_remove:
                if file_path.exists():
                    file_path.unlink()

            print("Client state reset successfully")
        except Exception as error:
            print(f"Failed to reset client state: {error}")
            raise error

    async def _register_client(self) -> None:
        """Register client and store new client ID"""
        try:
            # Clear existing client ID
            self.client_id = None

            # This would typically make an HTTP request to your MCP server's register endpoint
            # For this example, we'll simulate the registration
            registration_data = {
                "client_name": "MCP OAuth Client",
                "redirect_uris": ["http://localhost:3000/oauth/callback"],
                "scope": "read write"
            }

            # Simulate API call - replace with actual HTTP request
            # response = await self._make_registration_request(registration_data)
            response = {"client_id": f"mcp-client-{os.urandom(8).hex()}"}  # Simulated

            self.client_id = response["client_id"]

            # Cache the new client ID
            client_data = {
                "client_id": self.client_id,
                "created_at": asyncio.get_event_loop().time()
            }

            async with aiofiles.open(self.client_id_file, 'w') as f:
                await f.write(json.dumps(client_data, indent=2))

            print(f"Client registered successfully with ID: {self.client_id}")
        except Exception as error:
            print(f"Failed to register client: {error}")
            raise error

    async def initialize(self) -> None:
        """Initialize client with automatic state recovery"""
        try:
            # Try to load cached client ID
            await self._load_cached_state()

            # Initialize MCP session
            # This is a simplified example - adapt to your actual MCP setup
            async with stdio_client(["your-mcp-server"]) as (read, write):
                async with ClientSession(read, write) as session:
                    self.session = session
                    await session.initialize()

        except Exception as error:
            if self._is_invalid_client_id_error(error):
                print("Cached client ID invalid. Resetting and re-registering...")
                await self._reset_client_state()
                await self._register_client()
                # Retry initialization
                await self.initialize()
            else:
                raise error

    async def _load_cached_state(self) -> None:
        """Load cached client state"""
        try:
            if self.client_id_file.exists():
                async with aiofiles.open(self.client_id_file, 'r') as f:
                    data = json.loads(await f.read())
                    self.client_id = data["client_id"]

                    # Optional: check expiration (24 hours example)
                    # created_at = data.get("created_at", 0)
                    # if asyncio.get_event_loop().time() - created_at > 24 * 60 * 60:
                    #     raise ValueError("Client ID expired")

        except (FileNotFoundError, json.JSONDecodeError):
            # No cached state, will register new client
            pass

    async def _make_registration_request(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Make actual HTTP request to register client - implement based on your server"""
        # Example using aiohttp:
        # async with aiohttp.ClientSession() as session:
        #     async with session.post("https://your-mcp-server/register", json=data) as response:
        #         return await response.json()
        raise NotImplementedError("Implement actual registration API call")


# Usage example
async def create_mcp_client():
    client = OAuthMCPClient(cache_dir="./.mcp-cache")

    try:
        await client.initialize()
        print("MCP client initialized successfully")
        return client
    except Exception as error:
        print(f"Failed to initialize MCP client: {error}")
        raise error


if __name__ == "__main__":
    asyncio.run(create_mcp_client())